File: sdk/dotnet/fsharp/src/ModelBinding/SignalsModelBinder.fs
namespace StarFederation.Datastar.ModelBinding

open System.Text.Json.Nodes
open Microsoft.AspNetCore.Mvc.ModelBinding
open Microsoft.AspNetCore.Mvc.ModelBinding.Binders
open StarFederation.Datastar.DependencyInjection
open System.Text.Json
open StarFederation.Datastar.Utility

type SignalsModelBinder(signalsReader:IDatastarSignalsReaderService) =
    interface IModelBinder with
        member this.BindModelAsync(bindingContext) =
            let signalBindingContext = (bindingContext.BindingSource :?> DatastarSignalsBindingSource)
            match signalBindingContext.Path with
            | null | "" ->
                if bindingContext.ModelType.IsValueType || bindingContext.ModelType = typedefof<string> then
                    task {
                        let! signalsJson = signalsReader.ReadSignalsAsync()
                        try
                            let signalsJsonObject = JsonObject.Parse(signalsJson).AsObject()
                            let signals = JsonPath.getValue bindingContext.ModelType signalsJsonObject bindingContext.FieldName
                            bindingContext.Result <- ModelBindingResult.Success(signals)
                        with _ -> bindingContext.Result <- ModelBindingResult.Failed()
                    }
                else
                    task {
                        let! signalsJson = signalsReader.ReadSignalsAsync()
                        try
                            let signalsJsonObject = JsonObject.Parse(signalsJson).AsObject()
                            let signals =
                                if signalsJsonObject.ContainsKey bindingContext.FieldName
                                then JsonPath.getValue bindingContext.ModelType signalsJsonObject bindingContext.FieldName
                                else JsonSerializer.Deserialize(signalsJsonObject, bindingContext.ModelType)
                            bindingContext.Result <- ModelBindingResult.Success(signals)
                        with _ -> bindingContext.Result <- ModelBindingResult.Failed()
                    }
            | path ->
                task {
                    let! signalsJson = signalsReader.ReadSignalsAsync()
                    try
                        let signalsJsonObject = JsonObject.Parse(signalsJson).AsObject()
                        let signals = JsonPath.getValue bindingContext.ModelType signalsJsonObject path
                        bindingContext.Result <- ModelBindingResult.Success(signals)
                    with _ -> bindingContext.Result <- ModelBindingResult.Failed()
                }

type SignalsModelBinderProvider() =
    interface IModelBinderProvider with
        member this.GetBinder providerContext =
            let isSignalsBindingSource = true
            let isSignalsBindingSource = isSignalsBindingSource && providerContext <> null
            let isSignalsBindingSource = isSignalsBindingSource && providerContext.BindingInfo <> null
            let isSignalsBindingSource = isSignalsBindingSource && providerContext.BindingInfo.BindingSource <> null
            let isSignalsBindingSource = isSignalsBindingSource && providerContext.BindingInfo.BindingSource.DisplayName = DatastarSignalsBindingSource.BindingSourceName
            if isSignalsBindingSource
            then BinderTypeModelBinder(typedefof<SignalsModelBinder>)
            else null

File: sdk/dotnet/csharp/src/StarFederation.Datastar/ModelBinding/SignalsModelBinder.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.AspNetCore.Mvc.ModelBinding;
using StarFederation.Datastar.Interfaces;

namespace StarFederation.Datastar.ModelBinding;

/// <summary>
///     Model binder for Datastar signals.
/// </summary>
public class SignalsModelBinder : IModelBinder
{
    private readonly IDatastarSignalsReaderService _signalsReader;

    /// <summary>
    ///     Initializes a new instance of the <see cref="SignalsModelBinder" /> class.
    /// </summary>
    /// <param name="signalsReader">The signals reader service.</param>
    public SignalsModelBinder(IDatastarSignalsReaderService signalsReader)
    {
        _signalsReader = signalsReader;
    }

    /// <inheritdoc />
    public async Task BindModelAsync(ModelBindingContext? bindingContext)
    {
        if (bindingContext == null)
        {
            return;
        }

        var signalBindingContext = bindingContext.BindingSource as DatastarSignalsBindingSource;
        var path = signalBindingContext?.Path;

        if (string.IsNullOrEmpty(path))
        {
            if (bindingContext.ModelType.IsValueType || bindingContext.ModelType == typeof(string))
            {
                var signals = await _signalsReader.ReadSignalsAsStringAsync();
                try
                {
                    if (signals == null)
                    {
                        bindingContext.Result = ModelBindingResult.Failed();
                        return;
                    }

                    var signalsJsonObject = JsonNode.Parse(signals)!.AsObject();
                    var value = Utils.JsonPath.GetValue(bindingContext.ModelType, signalsJsonObject, bindingContext.FieldName);
                    bindingContext.Result = ModelBindingResult.Success(value);
                }
                catch
                {
                    bindingContext.Result = ModelBindingResult.Failed();
                }
            }
            else
            {
                var signals = await _signalsReader.ReadSignalsAsStringAsync();
                try
                {
                    if (signals == null)
                    {
                        bindingContext.Result = ModelBindingResult.Failed();
                        return;
                    }

                    var signalsJsonObject = JsonNode.Parse(signals)!.AsObject();
                    object? value;
                    if (signalsJsonObject.ContainsKey(bindingContext.FieldName))
                    {
                        value = Utils.JsonPath.GetValue(bindingContext.ModelType, signalsJsonObject, bindingContext.FieldName);
                    }
                    else
                    {
                        value = signalsJsonObject.Deserialize(bindingContext.ModelType);
                    }

                    bindingContext.Result = ModelBindingResult.Success(value);
                }
                catch
                {
                    bindingContext.Result = ModelBindingResult.Failed();
                }
            }
        }
        else
        {
            var signals = await _signalsReader.ReadSignalsAsStringAsync();
            try
            {
                if (signals == null)
                {
                    bindingContext.Result = ModelBindingResult.Failed();
                    return;
                }

                var signalsJsonObject = JsonNode.Parse(signals)!.AsObject();
                var value = Utils.JsonPath.GetValue(bindingContext.ModelType, signalsJsonObject, path);
                bindingContext.Result = ModelBindingResult.Success(value);
            }
            catch
            {
                bindingContext.Result = ModelBindingResult.Failed();
            }
        }
    }
}